### PR TITLE
Fix JSON array output

### PR DESCRIPTION
**Pull request by peterschrammel**
Fixes https://github.com/diffblue/platform/issues/1196, https://github.com/diffblue/test-gen/issues/21

### CODE REVIEW

File: src/goto-symex/symex_builtin_functions.cpp [hunk 1]
```diff
@@ -384,7 +390,7 @@ void goto_symext::symex_cpp_new(
              "dynamic_"+count_string+"_value";
   symbol.name="symex_dynamic::"+id2string(symbol.base_name);
   symbol.is_lvalue=true;
-  symbol.mode=ID_cpp;
+  symbol.mode=(code.get(ID_statement)==ID_java_new_array ? ID_java : ID_cpp);
```
Comment (allredj) [inline src/goto-symex/symex_builtin_functions.cpp]: Are we sure `ID_java_new_array` is the only case for a java statement? I'd rather we wrote something like this to make the code more robust:
```cpp
irep_idt &statement=code.get(ID_statement)
if(statement==ID_java_new_array)
  symbol.mode=ID_java;
else if(statement==ID_cpp_new_array)
  symbol.mode=ID_cpp;
else
  assert(false); // replace with the appropriate invariant

File: src/goto-programs/json_goto_trace.cpp [hunk 2]
```diff
@@ -117,8 +88,10 @@ exprt simplify_array_access(const exprt &src)
     remove_pointer_offsets(index);
 
     // We look for an actual integer value for the index
-    std::size_t i;
-    if(!simplify_index(index, i))
+    index=simplify_expr(index, ns);
```
Comment (peterschrammel) [inline src/goto-programs/json_goto_trace.cpp]: ?

File: src/goto-symex/symex_builtin_functions.cpp [hunk 3]
```diff
@@ -361,6 +361,11 @@ void goto_symext::symex_output(
   target.output(state.guard.as_expr(), state.source, output_id, args);
 }
 
+/// Handles side effects of type 'new' for C++ and 'new array'
+/// for C++ and Java language modes
+/// \param state: Symex state
+/// \param lhs: left-hand side of assignment
+/// \param code: right-hand side containing side effect
 void goto_symext::symex_cpp_new(
```
Comment (peterschrammel) [inline src/goto-symex/symex_builtin_functions.cpp]: Yes, definitely. Ideally, language-specific bits in goto-symex that are actually should be general should be refactored, but I suggest to do that in a separate PR.

File: src/goto-symex/symex_builtin_functions.cpp (hunk 1 above)
Comment (peterschrammel) [inline src/goto-symex/symex_builtin_functions.cpp]: fixed

File: src/goto-programs/json_goto_trace.cpp (hunk 2 above)
Comment (allredj) [inline src/goto-programs/json_goto_trace.cpp]: It's part of the `if`, so it should be moved two spaces to the right.

**Comment by cesaro** [comment]
I confirm that this fixes diffblue/test-gen#21 when tested on

- com/diffblue/javatest/BubbleSort.java: checkSort
- com/diffblue/javatest/Search.java: contains

File: regression/cbmc-java/json_trace3/test.desc [hunk 4]
```diff
@@ -0,0 +1,9 @@
+CORE
+Test.class
+--cover location --trace --json-ui --unwind 5
+activate-multi-line-match
+EXIT=0
+SIGNAL=0
+"lhs": "dynamic_2_array\[1L\]",\n *"mode": "java",\n *"sourceLocation": \{\n *"bytecodeIndex": "15",\n *"file": "Test\.java",\n *"function": "java::Test\.main:\(\[J\)V",\n *"line": "6"\n *\},\n *"stepType": "assignment",\n *"thread": 0,\n *"value": \{\n *"binary": "0000000000000000000000000000000000000000000000000000000000000001",\n *"data": "1L",\n *"name": "integer",\n *"type": "long",\n *"width": 64
+--
```
Comment (cesaro) [inline regression/cbmc-java/json_trace3/test.desc]: I propose to add here

```
"name": "unknown"
```

File: src/goto-programs/json_goto_trace.cpp (hunk 2 above)
Comment (cesaro) [inline src/goto-programs/json_goto_trace.cpp]: Isn't the `if` a removed line?

Comment (allredj) [inline src/goto-programs/json_goto_trace.cpp]: Oh I'm sorry.................. I need to book a colour-blindness test...

File: regression/cbmc-java/json_trace3/test.desc [hunk 5]
```diff
@@ -0,0 +1,10 @@
+CORE
+Test.class
+--cover location --trace --json-ui --unwind 5
```
Comment (allredj) [inline regression/cbmc-java/json_trace3/test.desc]: We generally try to use the same options as no platform, but it is not a strict rule.

Comment (peterschrammel) [inline regression/cbmc-java/json_trace3/test.desc]: This is CBMC, not test-gen.

Comment (allredj) [inline regression/cbmc-java/json_trace3/test.desc]: Oh yes.
